addons/point_of_sale/models/pos_config.py

```diff
@@ -735,29 +735,41 @@ def setup_invoice_journal(self, company):
             else:
                 pos_config.write({'module_account': False})
 
-    def get_limited_products_loading(self):
-        self.env.cr.execute("""
-            WITH pm AS
-            (
-                     SELECT   product_id,
-                              Max(write_date) date
-                     FROM     stock_quant
-                     GROUP BY product_id)
-            SELECT    p.id
-            FROM      product_product p
-            LEFT JOIN product_template t
-            ON        (
-                                product_tmpl_id=t.id)
-            LEFT JOIN pm
-            ON        (
-                                p.id=pm.product_id)
-            WHERE     p.active
-            AND       t.available_in_pos
-            ORDER BY  t.priority DESC,
+    def get_limited_products_loading(self, fields, offset):
+        query = """
+            WITH pm AS (
+                  SELECT product_id,
+                         Max(write_date) date
+                    FROM stock_quant
+                GROUP BY product_id
+            )
+               SELECT p.id
+                 FROM product_product p
+            LEFT JOIN product_template t ON product_tmpl_id=t.id
+            LEFT JOIN pm ON p.id=pm.product_id
+                WHERE (
+                        t.available_in_pos
+                    AND t.sale_ok
+                    AND (t.company_id=%(company_id)s OR t.company_id IS NULL)
+                    AND %(available_categ_ids)s IS NULL OR t.pos_categ_id=ANY(%(available_categ_ids)s)
```

Comment:
I believe the expression on right of `OR` should be `t.pos_categ_id IN %(available_categ_ids)s`? `ANY` returns a boolean which should not be compared to the id.

Reply:
It's the same: https://stackoverflow.com/questions/34627026/in-vs-any-operator-in-postgresql
The reason why I used `ANY` here is because `available_categ_ids` is optional. This lead to two different possible cases:
- `available_categ_ids` is either `None` or a list of values
- `available_categ_ids` is either `[]` or a list of values
By using the `IN` operator, the first case doesn't work in case `available_categ_ids` is `None. Only the second case would work but not the optional part of the query `%(available_categ_ids)s IS NULL` which can roughly be translated to `[] is NULL` and this doesn't work.
`ANY` was the working solution that I found

Reply:
Woah! 🤯  
For our use-case, the stackoverflow answer also mentioned we can use this construct:
![Screenshot 2021-09-13 at 11 27 21](https://user-images.githubusercontent.com/3245568/133059612-93c08cc0-51e6-4efa-b146-fb0c6375f2b6.png)

Reply:
It is suggesting that the your original AND condition about `available_categ_ids`
```sql
AND %(available_categ_ids)s IS NULL OR t.pos_categ_id=ANY(%(available_categ_ids)s)
```
is the same as this:
```sql
AND (t.pos_categ_id=ANY(%(available_categ_ids)s)) IS NOT TRUE
```
But damn, the above expression is hard to read as hell.